### ISSUE TITLE
Callback swallowed leading to Mutex unlock error

### ISSUE THREAD

**Issue by 1j01** [jvilk/BrowserFS#221]
Whew! This took some debugging.

I got `Uncaught Error: unlock of a non-locked mutex`, and at first I thought it was the `readdir` failing (and it kind of is), then I noticed that I had typo'd "`file`" instead of another variable, and I was trying to hunt down where that variable came from, assuming it was some sort of getter that was causing the error, but it turned out that any error in my callback would cause the mutex error.
So I continued debugging, and here's what I found:

- It's swallowing errors when it should check for ENOENT, in OverlayFS's `readdir`, [here](https://github.com/jvilk/BrowserFS/blob/7ecbab597fd56a172f93c2e0d7dc97d12a77b101/src/backend/OverlayFS.ts#L652) (unrelatedly it should probably also check for ENOENT [here](https://github.com/jvilk/BrowserFS/blob/7ecbab597fd56a172f93c2e0d7dc97d12a77b101/src/backend/OverlayFS.ts#L856))
- It's wrapping my callback in `XmlHttpRequest.prototype.readdir`, and then
- It's calling my callback twice! the second time to handle the error from the first callback!

https://github.com/jvilk/BrowserFS/blob/73e4eca717f9e8c21836c5d3affb14dcfde1d9e3/src/backend/HTTPRequest.ts#L341-L347

The combined result of this is that swallows any errors in my callback, and gives that mutex error instead of any reasonable error.

I'll send a PR for this.

**Comment by jvilk** [comment]
Good catch! I should do an audit of the codebase, looking for potential re-applications of callbacks.

Is the OverlayFS issue that stat is swallowing an error code that it should return?

The stat code is doing a `mkdir -p` to find the parent directory that exists before creating all needed subdirectories; what error code is it swallowing that you'd expect it to return?

Similarly, `readdir` may operate on directories that only exist in `writable`, so errors from `readable.readdir` are expected. Is there a particular error code it's swallowing that it should return instead? Note that this code may also receive a `ENOTDIR` error if `readable` contains a file that was later deleted, so it's a bit more subtle than just checking for `ENOENT`.

**Comment by 1j01** [comment]
> Is the OverlayFS issue that stat is swallowing an error code that it should return?

I don't think the statDone one affected me, I was only using readdir (I haven't created any directories yet)

> what error code is it swallowing that you'd expect it to return?

Possibly EPERM, or an unexpected type of error like ended up getting passed to the callback in readdir (which should be an ApiError) `this._readable.readdir(p, (err: ApiError, rFiles: string[]) => {`
But I guess maybe not.

The main thing was the double callback, and the callback being called as part of a try block. I'll send a PR for that at least.

> I should do an audit of the codebase, looking for potential re-applications of callbacks.

I've started doing this audit now, and there sure are a lot of cases, but I'm on a roll. 😄

**Comment by jvilk** [comment]
Ignoring EPERM should be fine; it could be a directory that you didn't have permission to, a chmod caused it to move to writable (because you updated the directory/file), and now you do have access to it. Keep in mind readable is a read-only snapshot of some file system, and writable stores the changes. Data in readable may be out-of-sync with reality.

**Comment by 1j01** [comment]
Yeah, that makes sense.

**Comment by dr-vortex** [comment]
@1j01 What is the status of this issue? I've made a ton of improvements to BFS in recent months so this issue may not be relevant or present anymore.

**Comment by dr-vortex** [comment]
Please use https://github.com/browser-fs/core/issues/14
